posthog/models/sessions/sql.py

```diff
@@ -0,0 +1,141 @@
+    event_count_map AggregateFunction(sumMap, Tuple(Array(String), Array(Int64))),
+    -- duplicate the event count as a specific column for pageviews and autocaptures,
+    -- as these are used in some key queries and need to be fast
+    pageview_count SimpleAggregateFunction(sum, Int64),
+    autocapture_count SimpleAggregateFunction(sum, Int64),
+) ENGINE = {engine}
+"""
+
+SESSIONS_DATA_TABLE_ENGINE = lambda: AggregatingMergeTree(TABLE_BASE_NAME, replication_scheme=ReplicationScheme.SHARDED)
+
+SESSIONS_TABLE_SQL = lambda: (
+    SESSIONS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(min_first_timestamp)
+    -- order by is used by the aggregating merge tree engine to
+    -- identify candidates to merge, e.g. toDate(min_first_timestamp)
+    -- would mean we would have one row per day per session_id
+    -- if CH could completely merge to match the order by
+    -- it is also used to organise data to make queries faster
+    -- we want the fewest rows possible but also the fastest queries
+    -- since we query by date and not by time
+    -- and order by must be in order of increasing cardinality
+    -- so we order by date first, then team_id, then session_id
+    -- hopefully, this is a good balance between the two
+    ORDER BY (toDate(min_first_timestamp), team_id, session_id)
+SETTINGS index_granularity=512
+"""
+).format(
+    table_name=SESSIONS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=SESSIONS_DATA_TABLE_ENGINE(),
+)
+
+SESSIONS_TABLE_MV_SQL = (
+    lambda: """
+CREATE MATERIALIZED VIEW IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+TO {database}.{target_table}
+AS SELECT
+
+`$session_id` as session_id,
+team_id,
+
+-- it doesn't matter which distinct_id gets picked (it'll be somewhat random) as they can all join to the right person
+any(distinct_id) as distinct_id,
+
+min(timestamp) AS min_first_timestamp,
+max(timestamp) AS max_last_timestamp,
+
+groupUniqArray(JSONExtractString(properties, '$current_url')) AS urls,
+argMinState(JSONExtractString(properties, '$current_url'), timestamp) as entry_url,
+argMaxState(JSONExtractString(properties, '$current_url'), timestamp) as exit_url,
+
+argMinState(JSONExtractString(properties, '$referring_domain'), timestamp) as initial_referring_domain,
+argMinState(JSONExtractString(properties, 'utm_source'), timestamp) as initial_utm_source,
+argMinState(JSONExtractString(properties, 'utm_campaign'), timestamp) as initial_utm_campaign,
+argMinState(JSONExtractString(properties, 'utm_medium'), timestamp) as initial_utm_medium,
+argMinState(JSONExtractString(properties, 'utm_term'), timestamp) as initial_utm_term,
+argMinState(JSONExtractString(properties, 'utm_content'), timestamp) as initial_utm_content,
+argMinState(JSONExtractString(properties, 'gclid'), timestamp) as initial_gclid,
+argMinState(JSONExtractString(properties, 'gad_source'), timestamp) as initial_gad_source,
+
+sumMapState(([event], [toInt64(1)])) as event_count_map,
```

Comment:
It's Array<key>:Array<string>, and the arrays need to be the same length, it's how you can define a map with multiple keys/values

Reply:
ah, some CH magic... I had to use `groupUniqArrayArray` instead of `groupUniqAray` in a similar situation the other day

the more I learn anout CH, the more I know I don't know 🤣

Reply:
Yeah agreed